### ISSUE TITLE
encoding/xml: XML namespace prefixes not fully supported

### ISSUE THREAD

**Issue by gopherbot** [golang/go#6800]
by **anthonyeden**:

<pre>It is possible to create a prefix on elements by simply defining the prefix as part of
the name in the XMLName declaration, but the xmlns attribute is output without the
prefix. Example: <a href="http://play.golang.org/p/KeIHB4_z5L">http://play.golang.org/p/KeIHB4_z5L</a>

Thus it would seem that the encoding/xml package can only deal with document-level
namespaces, not nested namespaces, is this correct?</pre>

**Comment by rogpeppe** [comment]
Fixed by https://go-review.googlesource.com/2660

**Comment by jordan2175** [comment]
Can we verify that it will in fact work with an XML Namespace aspect like (note the various XMLNS prefixes “stix:, indicator:, cybox:”):

<stix:STIX_Package xsi:schemaLocation="
	http://cybox.mitre.org/common-2 <http://cybox.mitre.org/common-2> http://cybox.mitre.org/XMLSchema/common/2.1/cybox_common.xsd
	http://cybox.mitre.org/cybox-2 http://cybox.mitre.org/XMLSchema/core/2.1/cybox_core.xsd
	http://cybox.mitre.org/default_vocabularies-2
	http://cybox.mitre.org/XMLSchema/default_vocabularies/2.1/cybox_default_vocabularies.xsd
	http://cybox.mitre.org/objects#URIObject-2 http://cybox.mitre.org/XMLSchema/objects/URI/2.1/URI_Object.xsd
	http://stix.mitre.org/Indicator-2 http://stix.mitre.org/XMLSchema/indicator/2.1.1/indicator.xsd
	http://stix.mitre.org/common-1 http://stix.mitre.org/XMLSchema/common/1.1.1/stix_common.xsd
	http://stix.mitre.org/default_vocabularies-1 http://stix.mitre.org/XMLSchema/default_vocabularies/1.1.1/stix_default_vocabularies.xsd
	http://stix.mitre.org/stix-1 http://stix.mitre.org/XMLSchema/core/1.1.1/stix_core.xsd"
	id="example:Package-8fab937e-b694-11e3-b71c-0800271e87d2"
	timestamp="2014-05-08T09:00:00.000000Z"
	version="1.1.1">

	<stix:Indicators>
		<stix:Indicator id="example:Indicator-d81f86b9-975b-bc0b-775e-810c5ad45a4f" xsi:type="indicator:IndicatorType">
			<indicator:Title>Malicious site hosting downloader</indicator:Title>
			<indicator:Type xsi:type="stixVocabs:IndicatorTypeVocab-1.0">URL Watchlist</indicator:Type>
			<indicator:Observable id="example:Observable-ee59c28e-d922-480e-9b7b-a79502696505">
				<cybox:Object id="example:URI-b13ae3fc-80af-49c2-9de9-f713abc070ba">
					<cybox:Properties xsi:type="URIObj:URIObjectType" type="URL">
						<URIObj:Value condition="Equals">http://x4z9arb.cn/4712</URIObj:Value>
					</cybox:Properties>
				</cybox:Object>
			</indicator:Observable>
		</stix:Indicator>
	</stix:Indicators>
</stix:STIX_Package>


Thanks,
Bret
PGP Fingerprint: 62A6 5999 0F7D 0D61 4C66 D59C 2DB5 111D 63BC A303
"Without cryptography vihv vivc ce xhrnrw, however, the only thing that can not be unscrambled is an egg."

**Comment by rogpeppe** [comment]
I'm sorry, I don't quite understand what you mean when you ask whether "it will in fact work with an XML Namespace aspect like ...".

**Comment by jordan2175** [comment]
Given the XML example, can your patch parse it correctly and then re-marshall it back in to the same thing?  I just want to make sure that it works with all of the namespace prefixes.  Thanks SO MUCH for the patch and for fixing this issue.


Thanks,
Bret
PGP Fingerprint: 62A6 5999 0F7D 0D61 4C66 D59C 2DB5 111D 63BC A303
"Without cryptography vihv vivc ce xhrnrw, however, the only thing that can not be unscrambled is an egg."

**Comment by davecheney** [comment]
There are no xmlns attributes in the fragment you presented. That looks
like xml schema information.

**Comment by jordan2175** [comment]
Silly of me...  Here is an example with them.

<stix:STIX_Package
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:stix="http://stix.mitre.org/stix-1"
    xmlns:indicator="http://stix.mitre.org/Indicator-2"
    xmlns:cybox="http://cybox.mitre.org/cybox-2"
    xmlns:AddressObject="http://cybox.mitre.org/objects#AddressObject-2"
    xmlns:cyboxVocabs="http://cybox.mitre.org/default_vocabularies-2"
    xmlns:stixVocabs="http://stix.mitre.org/default_vocabularies-1"
    xmlns:ExampleNamespace="http://example.com/"
    xsi:schemaLocation="
    http://stix.mitre.org/stix-1 http://stix.mitre.org/XMLSchema/core/1.0.1/stix_core.xsd
    http://stix.mitre.org/Indicator-2 http://stix.mitre.org/XMLSchema/indicator/2.0.1/indicator.xsd
    http://cybox.mitre.org/default_vocabularies-2 http://cybox.mitre.org/XMLSchema/default_vocabularies/2.0.1/cybox_default_vocabularies.xsd
    http://stix.mitre.org/default_vocabularies-1 http://stix.mitre.org/XMLSchema/default_vocabularies/1.0.1/stix_default_vocabularies.xsd
    http://cybox.mitre.org/objects#AddressObject-2 http://cybox.mitre.org/XMLSchema/objects/Address/2.0.1/Address_Object.xsd"
    id="ExampleNamespace:STIXPackage-33fe3b22-0201-47cf-85d0-97c02164528d"
    version="1.0.1">
    <stix:STIX_Header>
        <stix:Title>Example watchlist that contains IP information.</stix:Title>
        <stix:Package_Intent xsi:type="stixVocabs:PackageIntentVocab-1.0">Indicators - Watchlist</stix:Package_Intent>
    </stix:STIX_Header>
    <stix:Indicators>
        <stix:Indicator xsi:type="indicator:IndicatorType" id="ExampleNamespace:Indicator-33fe3b22-0201-47cf-85d0-97c02164528d">
            <indicator:Type xsi:type="stixVocabs:IndicatorTypeVocab-1.0">IP Watchlist</indicator:Type>
            <indicator:Description>Sample IP Address Indicator for this watchlist. This contains one indicator with a set of three IP addresses in the watchlist.</indicator:Description>
            <indicator:Observable  id="ExampleNamespace:Observable-1c798262-a4cd-434d-a958-884d6980c459">
                <cybox:Object id="ExampleNamespace:Object-1980ce43-8e03-490b-863a-ea404d12242e">
                    <cybox:Properties xsi:type="AddressObject:AddressObjectType">
                        <AddressObject:Address_Value condition="Equals" apply_condition="ANY">10.0.0.0##comma##10.0.0.1##comma##10.0.0.2</AddressObject:Address_Value>
                    </cybox:Properties>
                </cybox:Object>
            </indicator:Observable>
        </stix:Indicator>
    </stix:Indicators>
</stix:STIX_Package>

**Comment by davecheney** [comment]
I'm sorry, github ate your upload. Can you put it in a file an attach or
use a pastebin service ?

**Comment by jordan2175** [comment]
I see that.  I edited the post above to fix it.  I also copied it to pastebin,here 
http://pastebin.com/kbtDUWdE

**Comment by jordan2175** [comment]
Here is the example from their website as well.

http://stix.mitre.org/about/example_watchlist.html <http://stix.mitre.org/about/example_watchlist.html>



Thanks,
Bret
PGP Fingerprint: 62A6 5999 0F7D 0D61 4C66 D59C 2DB5 111D 63BC A303
"Without cryptography vihv vivc ce xhrnrw, however, the only thing that can not be unscrambled is an egg."

**Comment by c4milo** [comment]
@rogpeppe, did this fix land in Go 1.4.x?

**Comment by c4milo** [comment]
@rogpeppe, I'm having problems consuming a webservice that is expecting namespaces declared at the top along with prefixes, and each inner element using the prefix as opposed to the actual namespace repeated each time. You can find more details about this issue here: https://github.com/hooklift/gowsdl/issues/17

**Comment by notzippy** [comment]
@c4milo This fix is not in go 1.4.1

**Comment by jordan2175** [comment]
Here is another example that the current XML bindings can not process..  It is a bit different than the current issues, I think, or maybe it is related.  If I take out all but one of the xmlns attributes, then it works.

```
<taxii_11:Discovery_Response 
    xmlns:taxii="http://taxii.mitre.org/messages/taxii_xml_binding-1"
    xmlns:taxii_11="http://taxii.mitre.org/messages/taxii_xml_binding-1.1"
    xmlns:tdq="http://taxii.mitre.org/query/taxii_default_query-1" 
    message_id="32898" 
    in_response_to="1">
</taxii_11:Discovery_Response>
```

**Comment by wookoouk** [comment]
I am trying to parse (small snippet):

```
<w:document xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"></w:document>
```
but I get: error: expected element type <w:document> but have <document>

I could change
```
XMLName xml.Name `xml:"w:document"`
```
to
```
XMLName xml.Name `xml:"document"`
```

but that breaks everything else.

**Comment by c4milo** [comment]
@wookoouk this fix is going to land in Go 1.5. So, we are going to still see this failing in Go 1.4.x

**Comment by wookoouk** [comment]
Is it available in HEAD?

**Comment by c4milo** [comment]
@wookoouk yes, master branch, aka tip, should have it.

**Comment by wookoouk** [comment]
Cheers @c4milo

**Comment by notzippy** [comment]
I found that the master did not make much difference, I used another package https://github.com/jteeuwen/go-pkg-xmlx to read the xml, correct the name space, and output the xml

**Comment by jordan2175** [comment]
To everyone that is writing the code / patches and everyone that is contributing examples of things that are not working, THANKS!!!!   What a great community.  Lets make Golang the best!

**Comment by wookoouk** [comment]
yep, sadly the same for me:
```sh
go version devel +6ffed30 Wed Mar 18 15:14:37 2015 +0000 darwin/amd64
[xmd] go run xmd.go tomd test.docx                                                                       16:58:32  ☁  master ☂ ✖ ⚡ ✭
word/document.xml
error: expected element type <w:document> but have <document>%
```

**Comment by c4milo** [comment]
This issue should be re-opened then.

**Comment by mikioh** [comment]
Please open a new issue if necessary. The above additional information confuses the issue whether the applied fix generates a new issue (brakes existing code and we should revert the fix) or just not fixing undiscovered, already reported issues.

If you are saying that the applied change 2660 should be reverted for some reason, please write on this issue so.
If you are saying that the applied change 2660 does not fix your issue, please search existing issues and write additional information for contributors, or open a new issue if necessary.

Running issue filters with "encoding/xml in:title" still reports there're 19 open issues.

**Comment by jordan2175** [comment]
I do not believe this issue has actually been fixed with patch #2660.  Especially when you consider the fact that issues #9519 and #9775 were closed and folded in to this issue and neither one of them has been addressed.  That means until they are also fixed, this issue is not done and should be reopened.

**Comment by mikioh** [comment]
@jordan2175,

Fine, re-open both #9519 and #9775 instead. But please add more useful information for contributors in #9519 and #9775. As @rogpeppe said in above:
```
I'm sorry, I don't quite understand what you mean when you ask whether "it will in fact work with an XML Namespace aspect like ...".
```
